File: backend/src/main/java/com/portfolio/domain/Project.java
package com.portfolio.domain;

import com.portfolio.domain.enums.ProjectStatus;
import jakarta.persistence.*;
import lombok.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

/**
 * Representa um projeto exibido no portfolio, com CRUD completo
 * disponível para o Admin.
 */

@Entity
@Table(name = "project")
@EntityListeners(AuditingEntityListener.class)
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Project {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "title", nullable = false, length = 150)
    private String title;

    @Column(name = "short_description", nullable = false, length = 300)
    private String shortDescription;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Column(name = "repository_url", length = 300)
    private String repositoryUrl;

    @Column(name = "demo_url", length = 300)
    private String demoUrl;

    @Column(name = "image_url", length = 500)
    private String imageUrl;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 30)
    @Builder.Default
    private ProjectStatus status = ProjectStatus.COMPLETED;

    @Column(name = "featured", nullable = false)
    @Builder.Default
    private boolean featured = false;

    @Column(name = "display_order", nullable = false)
    @Builder.Default
    private int displayOrder = 0;

    @Column(name = "start_date")
    private LocalDate startDate;

    @Column(name = "end_date")
    private LocalDate endDate;

    @Builder.Default
    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "project_tech_stack", joinColumns = @JoinColumn(name = "project_id"))
    @Column(name = "technology", nullable = false, length = 60)
    private List<String> techStack = new ArrayList<>();

    /**
     * Galeria de screenshots do projeto, ordenada por displayOrder. A
     * primeira imagem é exibida como capa. {@code imageUrl} continua
     * existindo como fallback simples para quem não usa a galeria.
     */
    @Builder.Default
    @OneToMany(mappedBy = "project", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    @OrderBy("displayOrder ASC")
    private List<ProjectImage> images = new ArrayList<>();

    @CreatedDate
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @LastModifiedDate
    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;
}

File: frontend/src/types/project.ts
export type ProjectStatus = "IN_PROGRESS" | "COMPLETED" | "ARCHIVED";

export interface ProjectImage {
  id: number;
  url: string;
  displayOrder: number;
}

export interface Project {
  id: number;
  title: string;
  shortDescription: string;
  description: string;
  repositoryUrl?: string;
  demoUrl?: string;
  imageUrl?: string;
  status: ProjectStatus;
  featured: boolean;
  displayOrder: number;
  startDate?: string; // LocalDate ISO (yyyy-MM-dd)
  endDate?: string;
  techStack: string[];
  images: ProjectImage[];
  createdAt: string;
  updatedAt: string;
}

export interface ProjectPayload {
  title: string;
  shortDescription: string;
  description: string;
  repositoryUrl?: string;
  demoUrl?: string;
  imageUrl?: string;
  status: ProjectStatus;
  featured: boolean;
  displayOrder: number;
  startDate?: string;
  endDate?: string;
  techStack: string[];
}

export const PROJECT_STATUS_LABEL: Record<ProjectStatus, string> = {
  IN_PROGRESS: "Em andamento",
  COMPLETED: "Concluído",
  ARCHIVED: "Arquivado",
};
